torch/distributed/_spmd/gm_transformation.py

```diff
@@ -0,0 +1,131 @@
+import operator
+from typing import Any, Callable, Dict, List, Optional
+
+import torch
+import torch.nn as nn
+
+from functorch import make_fx
+from torch import fx
+from torch._inductor.compile_fx import compile_fx_inner
+from torch._inductor.decomposition import select_decomp_table
+from torch.distributed._spmd.graph_utils import OP
+from torch.distributed._spmd.iter_graph_module import IterGraphModule
+from torch.utils._pytree import tree_flatten, tree_map_only
+
+
+class InductorWrapper(nn.Module):
+    def __init__(self, gm: fx.GraphModule, enable_cudagraphs: bool) -> None:
+        super().__init__()
+        self._gm = gm
+        self._compiled: Optional[nn.Module] = None
+        self._enable_cudagraphs = enable_cudagraphs
+
+    def forward(self, *args: Any) -> Any:
+        if self._compiled is None:
+            gm = make_fx(self._gm, decomposition_table=select_decomp_table())(*args)
+            self._compiled = compile_fx_inner(
+                gm,
+                list(args),
+                cudagraphs=self._enable_cudagraphs,
+            )
+        list_args, _ = tree_flatten(args)
+        return self._compiled(list_args)
+
+
+def lower_to_inductor(
+    gm: torch.fx.GraphModule, enable_cudagraphs: bool
+) -> torch.fx.GraphModule:
+    orig_placeholders = []
+    orig_output_args: List[Any] = []
+    output = next(iter(gm.graph.nodes))
+    move_nodes = []
+
+    for node in gm.graph.nodes:
+        if node.op == OP.OUTPUT:
+            output = node
+            orig_output_args, _ = tree_flatten(node.args)
+        elif node.op == OP.PLACEHOLDER:
+            orig_placeholders.append(node)
+        else:
+            move_nodes.append(node)
+    orig_placeholders_set = set(orig_placeholders)
+
+    subgraph: torch.fx.Graph = torch.fx.Graph()
+    node_mapping: Dict[torch.fx.Node, torch.fx.Node] = {}
+    attrs = {}
+    for i, p in enumerate(orig_placeholders):
+        node_mapping[p] = subgraph.placeholder(name=f"placeholder_{i}")
+
+    for node in move_nodes:
+        if node.op == OP.GET_ATTR:
+            attrs[node.target] = getattr(gm, node.target)
+        args, kwargs = tree_map_only(
+            fx.Node, lambda n: node_mapping[n], (node.args, node.kwargs)
+        )
+        node_mapping[node] = subgraph.create_node(
+            op=node.op, target=node.target, args=args, kwargs=kwargs
+        )
+        node_mapping[node].meta = node.meta
```

Comment:
`node_copy` seems handles both arg remap and .meta field

Reply:
good point. We should use `node_copy` here.